test/transfer-manager.ts

```diff
@@ -0,0 +1,294 @@
+        ws.emit('complete');
+        ws.end();
+        return true;
+      };
+      return ws;
+    };
+  }
+}
+
+class HTTPError extends Error {
+  code: number;
+  constructor(message: string, code: number) {
+    super(message);
+    this.code = code;
+  }
+}
+
+let pLimitOverride: Function | null;
+const fakePLimit = (limit: number) => (pLimitOverride || pLimit)(limit);
+const fakeFs = extend(true, {}, fs, {
+  get promises() {
+    return {
+      open: () => {
+        return {
+          close: () => {},
+          write: (buffer: Buffer) => {
+            return Promise.resolve({buffer});
+          },
+        };
+      },
+      lstat: () => {
+        return {
+          isDirectory: () => {
+            return false;
+          },
+        };
+      },
+    };
+  },
+});
+
+describe('Transfer Manager', () => {
+  let TransferManager: any;
+  let transferManager: any;
+  let Bucket: any;
+  let bucket: any;
+  let File: any;
+
+  const STORAGE: any = {
+    createBucket: util.noop,
+    retryOptions: {
+      autoRetry: true,
+      maxRetries: 3,
+      retryDelayMultipier: 2,
+      totalTimeout: 600,
+      maxRetryDelay: 60,
+      retryableErrorFn: (err: HTTPError) => {
+        return err.code === 500;
+      },
+      idempotencyStrategy: IdempotencyStrategy.RetryConditional,
+    },
+    crc32cGenerator: () => new CRC32C(),
+  };
+  const BUCKET_NAME = 'test-bucket';
+
+  before(() => {
+    Bucket = proxyquire('../src/bucket.js', {
+      'p-limit': fakePLimit,
+      './nodejs-common': {
+        ServiceObject: FakeServiceObject,
+        util: fakeUtil,
+      },
+      './acl.js': {Acl: FakeAcl},
+      './file.js': {File: FakeFile},
+    }).Bucket;
+
+    File = proxyquire('../src/file.js', {
+      './nodejs-common': {
+        ServiceObject: FakeServiceObject,
+        util: fakeUtil,
+      },
+    }).File;
+
+    TransferManager = proxyquire('../src/transfer-manager.js', {
+      'p-limit': fakePLimit,
+      './nodejs-common': {
+        ServiceObject: FakeServiceObject,
+        util: fakeUtil,
+      },
+      './acl.js': {Acl: FakeAcl},
+      './file.js': {File: FakeFile},
+      fs: fakeFs,
+      fsp: fakeFs,
+    }).TransferManager;
+  });
+
+  beforeEach(() => {
+    bucket = new Bucket(STORAGE, BUCKET_NAME);
+    transferManager = new TransferManager(bucket);
+  });
+
+  describe('instantiation', () => {
+    it('should correctly set the bucket', () => {
+      assert.strictEqual(transferManager.bucket, bucket);
+    });
+  });
+
+  describe('uploadManyFiles', () => {
+    it('calls upload with the provided file paths', async () => {
+      const paths = ['/a/b/c', '/d/e/f', '/h/i/j'];
+      let count = 0;
+
+      bucket.upload = (path: string) => {
+        count++;
+        assert(paths.includes(path));
+      };
+
+      await transferManager.uploadManyFiles(paths);
+      assert.strictEqual(count, paths.length);
+    });
+
+    it('sets ifGenerationMatch to 0 if skipIfExists is set', async () => {
```

Comment:
is there a test for `skipIfExists` set to true, the file exists, and it errors? We should verify that it doesnt stop the execution of the code and just moves onto the next file.